### ISSUE TITLE
Questions about the dimming.ts file

### ISSUE THREAD

**Issue by MoTechnicalities** [kpsuperplane/homebridge-wiz-lan#92]
@kpsuperplane

I am trying to help debug.
RE: [https://github.com/kpsuperplane/homebridge-wiz-lan/blob/master/src/accessories/WizBulb/characteristics/dimming.ts](url)
I plugged the essence of the following functions from your dimming.ts file into an excel spreadsheet and I think they produce inappropriate results for the Characteristic.Brightness.

Line 16:
`return Number(Math.round((Math.max(10, Number(pilot.dimming)) - 100) * 1.1 + 100));`

Line 44: 
`{ dimming: Math.round((Math.max(1, Number(newValue)) + 10) / 1.1) },`

The **Characteristic.Brightness** is supposed to be an integer value  from 0 to 100 with a minimum step value of 1.
In the current setup, neither Line 16 nor Line 44 allow for a value of zero.
see [https://developers.homebridge.io/#/characteristic/Brightness](url)

**Comment by MoTechnicalities** [comment]
I would recommend using Math.floor function to derive the needed integer value, as it returns the largest integer less than or equal to a given number.

Using Math.floor(), an input Number range of 0 <--> 100.99 would yield output integer values of 0 <--> 100.

**Comment by xmanu** [comment]
The issue you describe in #88 seems to stem from something different, as NaN is passed as brightness (not 0), at least that is what the log message complains about.

The floor function would not suffice for your proposed fix. The issue I solved initially is that Wiz Lights don't accept values from 1 to 9 as brightness.
So the code you mentioned above will "normalize" brightness from 1 to 100 to the range of 10 to 100 accepted by the wiz lights. This code was written before the scene support was added.
Without scene support I can set up automations that turn off lights without issues. I guess that there is some interaction between the new scene feature and automations that makes this problematic.

I agree though that the above code should accept 0 as a brightness value to be conformant with the homebridge documentation. When I find the time I will update the code to also accept 0.
Note though, that I cannot currently test this in conjunction with the scene feature, as my homebridge setup cannot enable the feature without breaking my setup (I have lots of grouped lights that by documentation don't work with scene support).

**Comment by MoTechnicalities** [comment]
> The issue you describe in #88 seems to stem from something different, as NaN is passed as brightness (not 0), at least that is what the log message complains about.
> 
NaN is passed because the function in Line 16 attempts to return unsupported floating point (non integer) numbers.

For example:
 if pilot.dimming = 32 then the function in its current form will return the unsupported value of 25.2

Homebridge will then provide a NaN (unsupported number) warning because the number is expected to be an integer, not a floating point for the feature.

**Comment by MoTechnicalities** [comment]
@xmanu 
Hi, please see #93 to see if that would work. Thanks.

**Comment by MoTechnicalities** [comment]
@xmanu

> The issue you describe in #88 seems to stem from something different, as NaN is passed as brightness (not 0), at least that is what the log message complains about.

You are correct. I tried adjusting the function in the above lines on my own system. 
I found specifically that NaN is passed when getPilot fails to receive dimmer information from a light. ( See my logs posted in Issue #84 ).

Additionally, adjusting the above functions has no effect on whatever bug is preventing us from setting a light to 0% dimming when attempting to create a scene ( Per Issue #84 ).

Thanks for your response.

**Comment by MoTechnicalities** [comment]
> Should be fixed in 3.2.1. Can anyone verify?

@kpsuperplane 
Hi Kevin,
The problem is NOT FIXED. (PLEASE READ ON!)
I have analyzed and tested the problem carefully. I now know what the problem is! But I do not know your program well enough to fix it for you.
**Here is the station:**
To be sure, Wiz bulbs utilize a "dimming" property. However, it does not relate AT ALL to HomeKit's "brightness" characteristic! It only relates to Wiz scenes! And, to complicate things further, neither the bulbs nor the Wiz app utilize brightness controls when the bulb is set to {"sceneId":6} (Nightlight mode), and when the bulb is in that mode you cannot adjust brightness using the Home App either. You would have to first adjust the color in the Home App before you can change the brightness.

So, I read up on the "brightness" characteristic in the HomeKit Accessory Protocol Specification manual. The manual states:
   "The value is expressed as a percentage (%) of the maximum level of supported brightness."

As it turns out "brightness" can be calculated using the "highest" RGBCW (Red, Green, Blue, Cool White, and Worm White) values.
**For example:** 
We know that the highest possible of any of these values is 255, range being 0 to 255. So if we have the following color values.
"r":35
"g":255
"b":140
"c":0
"w":213
We know that this is the highest possible value for that perceived color. You can't make that color brighter! Because Green is at the Max value of 255 for that color. The program only needs to track the perceived color.
So THEN at 45% brightness for that perceived color, the values will be calculated as follows:
"r":35 * 45% = 15
"g":255 * 45% = 114
"b":140 * 45% = 63
"c":0 * 45% = 0
"w":213 * 45% = 95
SAME COLOR but now at only 45%. The formula is [Math.floor(value * 45%)]
Turning the bulb off can also be set this way:
"r":35 * 0% = 0
"g":255 * 0% = 0
"b":140 * 0% = 0
"c":0 * 0% = 0
"w":213 * 0% = 0
 
To determine a current brightness in a perceived color, you need to take the individual color with the highest value, and calculate as follows:
Given set values for r,g,b,c, and w, (no matter what those values are), if RED has the highest value at 178, then "brightness" = [Math.ceiling(178 * 100)] = 128.

You only need to provide the bulb with specific r,g,b,c, and w values for both HomeKit and the bulb to be in sync with perceived color and brightness.
And if you want to use the bulb scenes, then you only need to send a "sceneId" and a "dimming" value (I would suggest "dimming":100, then let the user adjust), except for "sceneId":6, which does not use dimming values at all and generates the Homebridge NaN log error because of the way the plugin currently tries to deal with "dimming".
 
I hope that helps!

**Comment by MoTechnicalities** [comment]
What I'm saying is don't use Wiz bulb dimming value to set brightness. Obtain the bulbs r, g, b, c, and w values and use those to calculate and pass brightness value to HomeKit. And don't set the bulbs dimming from HomeKit brightness values. Instead calculate the r, g, b, c, and w values from HomeKit and setPilot those rgbcw those values to the bulb.

**Comment by catongates** [comment]
I thought that pr-99 addressed this issue, but as of today, with v3.2.5 of the plugin, I'm still seeing quite a lot of this in my logs:

> [1/14/2023, 12:21:00 PM] [homebridge-wiz-lan] This plugin generated a warning from the characteristic 'Brightness': characteristic value expected valid finite number and received "NaN" (number). See https://homebridge.io/w/JtMGR for more info.

I assume I'm not alone in this.
